cmd/k8s-operator/hostscm.go

```diff
@@ -0,0 +1,158 @@
+	targetIP string
+	// serviceLabels identify the proxy service
+	serviceLabels map[string]string
+}
+
+// hostsCMProvisioner knows how to update ts.net nameserver config as egress
+// proxy Services get added/deleted
+type hostsCMProvisioner struct {
+	// Client reads from cache first, then kube-apiserver, writes to
+	// kube-apiserver
+	client.Client
+	// namespace in which tailscale resources get provisioned
+	tsNamespace string
+	// localClient knows how to talk to tailscaled local API
+	localAPIClient localClient
+}
+
+type localClient interface {
+	WhoIs(ctx context.Context, remoteAddr string) (*apitype.WhoIsResponse, error)
+}
+
+// Provision ensures that ts.net nameserver config has been updated with a
+// mapping of Tailscale service to its egress proxy kube Service IP. If
+// successful, returns the Tailscale service FQDN
+func (h *hostsCMProvisioner) Provision(ctx context.Context, logger *zap.SugaredLogger, hcc *hostsCMConfig) (string, error) {
+	logger.Info("IN CM PROV")
+	ip := fmt.Sprintf("%s:0", hcc.targetIP)
+	whois, err := h.localAPIClient.WhoIs(ctx, ip)
+	if err != nil {
+		logger.Errorf("error determining tailscale service: %v", err)
+		return "", err
+	}
+	fqdn := whois.Node.Name
+	fqdn = strings.TrimSuffix(fqdn, ".")
+
+	proxySvc, err := getSingleObject[corev1.Service](ctx, h.Client, h.tsNamespace, hcc.serviceLabels)
+	if apierrors.IsNotFound(err) {
+		// we will reconcile again on proxy Service creation/update
+		// event and the hosts config will get updated then
+		logger.Debugf("proxy Service not yet created waiting...")
+		return "", nil
+	}
+	if err != nil {
+		logger.Errorf("error retrieving proxy Service: %v", err)
+		return "", err
+	}
+
+	if proxySvc == nil || proxySvc.Spec.ClusterIP == "" || proxySvc.Spec.ClusterIP == "None" {
+		// we will reconcile again on proxy Service creation/update
+		// event and the hosts config will get updated then
+		logger.Infof("proxy Service for %#+v not yet ready, waiting...", proxySvc)
+		return "", nil
+	}
+
+	logger.Debugf("ensuring a ts.net record for %s: %s", fqdn, proxySvc.Spec.ClusterIP)
+
+	cm := &corev1.ConfigMap{}
+	err = h.Get(ctx, types.NamespacedName{Name: dnsConfigMapName, Namespace: h.tsNamespace}, cm)
+	if err != nil {
+		logger.Errorf("error retrieving hosts config: %v", err)
+		return "", err
+	}
+
+	hosts := make(map[string]string)
+	if cm.Data[dnsConfigKey] != "" {
+		if err := json.Unmarshal([]byte(cm.Data[dnsConfigKey]), &hosts); err != nil {
+			logger.Errorf("error unmarshaling hosts config %v", err)
+			return "", err
+		}
+	}
+	hosts[fqdn] = proxySvc.Spec.ClusterIP
+	hostsBytes, err := json.Marshal(hosts)
+	if err != nil {
+		logger.Errorf("error marshaling hosts config %v", err)
+		return "", err
+	}
+
+	cm.Data[dnsConfigKey] = string(hostsBytes)
+
+	// TODO (irbekrm): probably better to SSA here
+	// TODO (irbekrm): check diff only apply update if needed
+	if err := h.Update(ctx, cm); err != nil {
```

Comment:
Just curious, what would happen if the ConfigMap was modified between our Get and this Update?

Reply:
The Update would fail because the `metadata.resourceVersion` would have changed in kube-apiserver see [docs](https://kubernetes.io/docs/reference/using-api/api-concepts/#:~:text=You%20can%20overwrite,Conflict%20error%20response.) which would cause this reconcile to exit with an error and be retried with backoff till it succeeds. We watch this configmap so we will eventually have the up to date version of the configmap in cache. In practice, it would not take more than one extra reconcile and actually we no-one else is expected to update this configmap.